### PR TITLE
Add CollectionSegment<T>

### LINKED ISSUE #3596: CollectionSegment<T> is missing
Offset pagination can't be easily extended when using the Annotation-based approach, since `CollectionSegment<T>` (counterpart of `Connection<T>`) has not been implemented. Since we only have a non generic `CollectionSegment` the return type can not be inferred. Therefor the following breaks:
```csharp
public class Query
{
    [UseOffsetPaging(typeof(User))]
    public CollectionSegment GetUsers(int? take, int? skip)
    {
        var collection = new CollectionSegment(_users, 
            new CollectionSegmentInfo(true, false), ct => ValueTask.FromResult(0));        
        
        return collection;
    }
}
```

A `CollectionSegment<T>` should be implemented so we can write:
```csharp
public class Query
{
    [UseOffsetPaging]
    public CollectionSegment<User> GetUsers(int? take, int? skip)
    {
        var collection = new CollectionSegment<User>(_users, 
            new CollectionSegmentInfo(true, false), ct => ValueTask.FromResult(0));        
        
        return collection;
    }
}
```

### PR DESCRIPTION
**Pull request by tobias-tengler**
Closes #3596

### CODE REVIEW

File: src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment.cs [hunk 1]
```diff
@@ -31,7 +31,8 @@ public class CollectionSegment : IPage
         {
             Items = items ??
                 throw new ArgumentNullException(nameof(items));
-            Info = info;
+            Info = info ??
+                throw new ArgumentNullException(nameof(info));
```
Comment (tobias-tengler) [inline src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment.cs:35]: This wasn't here before, but I think it should've.

File: src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment~1.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace HotChocolate.Types.Pagination
+{
+    /// <inheritdoc />
+    public class CollectionSegment<T> : CollectionSegment
+        where T : class
```
Comment (tobias-tengler) [inline src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment~1.cs:10]: @michaelstaib Is this fine? i didn't find a quicker solution for being able to pass it to the base constructor which needs an `IReadOnlyList<object>`.

File: src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment.cs (hunk 1 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment.cs:35]: Was I think also a mistake that it was not there.

Comment (tobias-tengler) [reply src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment.cs:35]: It's declared as non-nullable in the schema, so I think it's the right thing to do.

File: src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment~1.cs (hunk 2 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types.OffsetPagination/CollectionSegment~1.cs:10]: This aligns great with the connection types.

**Review by michaelstaib — APPROVED** [review]

**Pull request merged**